<a href="https://colab.research.google.com/github/vbhaktha13/Neural-Networks/blob/main/Low_Code_%26_Auto_ML_with_PyCaret_Kickstarter_Exercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Low-Code & Auto-ML with PyCaret - Kickstarter

Welcome to this hands-on workshop session where we will learn about leveraging the very popular low-code and auto-ml library PyCaret!

![](https://i.imgur.com/cWzC62x.png)


The focus of this notebook is more of a kickstarter notebook to get your feet wet and get familiar with PyCaret before we hop on to more complex things! 

# 1 - Objective


The objective of this notebook will be to tackle a classification problem using the power of PyCaret. We will be exploring the following features and capabilities of PyCaret along the way:

-  __Getting Data:__ Learn how to import default datasets from PyCaret repository

-  __Environment Setup:__ Learn how to setup an experiment in PyCaret and get started with classification tasks!

-  __Compare Models:__ Learn how to compare multiple machine learning models for the given classification task based on model evaluation metrics

-  __Create Model:__ Learn how to create specific classifical models, perform stratified cross validation and evaluate classification metrics

-  __Tune Model:__ Learn how to automatically tune the hyper-parameters of classification models

-  __Plot Model:__ Learn how to analyze model performance using various diagnostic plots

-  __Finalize Model:__ Learn how to finalize the best model at the end of the experiment

-  __Predict Model:__ Learn how to make predictions on new / unseen data 

-  __Save / Load Model:__ Learn how to save / load a model for future use

# 2 - Install PyCaret

The first step to get started is to install `pycaret`. 

Run the following two lines




In [ ]:
!pip install pycaret

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 320 kB 5.3 MB/s 
     |████████████████████████████████| 2.0 MB 43.4 MB/s 
     |████████████████████████████████| 1.7 MB 29.0 MB/s 
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
    Preparing wheel metadata ... done
     |████████████████████████████████| 17.9 MB 432 kB/s 
     |████████████████████████████████| 262 kB 70.3 MB/s 
     |████████████████████████████████| 88 kB 6.6 MB/s 
     |████████████████████████████████| 167 kB 54.3 MB/s 
     |████████████████████████████████| 1.3 MB 37.1 MB/s 
     |████████████████████████████████| 56 kB 4.1 MB/s 
     |████████████████████████████████| 6.8 MB 38.9 MB/s 
     |████████████████████████████████| 125 kB 42.3 MB/s 
     |████████████████████████████████| 10.4 MB 27.7 MB/s 
     |████████████████████████████████| 

In [ ]:
!pip install -U jinja2 # https://github.com/pycaret/pycaret/issues/2591 to bypass import errors later on

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 133 kB 5.2 MB/s 
  Attempting uninstall: jinja2
    Found existing installation: Jinja2 2.11.3
    Uninstalling Jinja2-2.11.3:
      Successfully uninstalled Jinja2-2.11.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires requests~=2.23.0, but you have requests 2.28.1 which is incompatible.
flask 1.1.4 requires Jinja2<3.0,>=2.10.1, but you have jinja2 3.1.2 which is incompatible.
datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.


## Restart the Kernel now and then proceed by running the following cells as usual

## 2.1 - Enable Interactive Visuals

If you are using Google Colab, please run the following to enable interactive visuals

In [ ]:
from pycaret.utils import enable_colab
enable_colab()

Colab mode enabled.


# 3 - Binary Classification

The objective in this notebook will be to solve a predictive machine learning classification problem. To be more specific, it is going to be binary classification.

Binary classification is a supervised machine learning technique where the key objective is to predict a response variable given a set of independent variables (features). The response variable is categorical, having two discrete class labels, such as 1/0, Yes/No, Positive/Negative, Default/Not-Default and so on. 

A few real world use cases for classification are listed below:

- Fraud detection models to detect if a transaction is fraudulent or not fraudulent
- A "pass or fail" test method or quality control in factories, i.e. deciding if a specification has or has not been met – a go/no-go classification.
- Sentiment Analysis -> Positive or Negative

# 4 - PyCaret Classification Module

PyCaret's `classification` module (`pycaret.classification`) is a supervised machine learning module which is used for training, tuning, evaluating and deploying classification models. 

The PyCaret `classification` module can be used for Binary or Multi-class classification problems. It has over 18 algorithms and 14 plots to analyze the performance of models. Be it hyper-parameter tuning, ensembling or advanced techniques like stacking, PyCaret's classification module has it all.

Do check out [`pycaret.classification`'s documentation](https://pycaret.gitbook.io/docs/get-started/quickstart#classification)and [full-fledged APIs](https://pycaret.readthedocs.io/en/latest/api/classification.html) as needed!

# 5 - Getting the Data

We will be using a popular open-source dataset, called the "Adult" dataset also known as "Census Income" dataset.

Key Objective: Predict whether income exceeds $50K/yr based on census data

You can download the data from the original source [found here](https://archive.ics.uci.edu/ml/datasets/adult) 


and load it using `pandas` or you can use PyCaret's data respository to load the data using the `get_data()` function (This will require an internet connection).

The PyCaret version of the dataset is slightly more processed and is a subset.

## 5.1 - Data Retrieval

In [ ]:
from pycaret.datasets import get_data
dataset = get_data('income')

age         workclass  education  education-num      marital-status  \
0   39         State-gov  Bachelors             13       Never-married   
1   50  Self-emp-not-inc  Bachelors             13  Married-civ-spouse   
2   38           Private    HS-grad              9            Divorced   
3   53           Private       11th              7  Married-civ-spouse   
4   28           Private  Bachelors             13  Married-civ-spouse   

          occupation   relationship   race     sex  capital-gain  \
0       Adm-clerical  Not-in-family  White    Male          2174   
1    Exec-managerial        Husband  White    Male             0   
2  Handlers-cleaners  Not-in-family  White    Male             0   
3  Handlers-cleaners        Husband  Black    Male             0   
4     Prof-specialty           Wife  Black  Female             0   

   capital-loss  hours-per-week native-country  income >50K  
0             0              40  United-States            0  
1             0              13  United-States            0  
2             0              40  United-States            0  
3             0              40  United-States            0  
4             0              40           Cuba            0

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       30725 non-null  object
 2   education       32561 non-null  object
 3   education-num   32561 non-null  int64 
 4   marital-status  32561 non-null  object
 5   occupation      30718 non-null  object
 6   relationship    32561 non-null  object
 7   race            32561 non-null  object
 8   sex             32561 non-null  object
 9   capital-gain    32561 non-null  int64 
 10  capital-loss    32561 non-null  int64 
 11  hours-per-week  32561 non-null  int64 
 12  native-country  31978 non-null  object
 13  income >50K     32561 non-null  int64 
dtypes: int64(6), object(8)
memory usage: 3.5+ MB


In [ ]:
dataset.shape

(32561, 14)

In [ ]:
dataset.iloc[25:30]

age  workclass     education  education-num      marital-status  \
25   56  Local-gov     Bachelors             13  Married-civ-spouse   
26   19    Private       HS-grad              9       Never-married   
27   54        NaN  Some-college             10  Married-civ-spouse   
28   39    Private       HS-grad              9            Divorced   
29   49    Private       HS-grad              9  Married-civ-spouse   

         occupation   relationship                race   sex  capital-gain  \
25     Tech-support        Husband               White  Male             0   
26     Craft-repair      Own-child               White  Male             0   
27              NaN        Husband  Asian-Pac-Islander  Male             0   
28  Exec-managerial  Not-in-family               White  Male             0   
29     Craft-repair        Husband               White  Male             0   

    capital-loss  hours-per-week native-country  income >50K  
25             0              40  United-States            1  
26             0              40  United-States            0  
27             0              60          South            1  
28             0              80  United-States            0  
29             0              40  United-States            0

## 5.2 - Split Data into Train-Test Datasets

In order to demonstrate the `predict_model()` function on unseen data, a holdout sample of 15% records has been withheld from the original dataset to be used for predictions. 

This will be your true unseen test dataset to be used at the end once all training is complete as a simulation of live real data.

In [ ]:
# create train - test datasets
data_train = dataset.sample(frac=0.85, random_state=42)
data_test = dataset.drop(data_train.index)

# reset row numbers \ indices
data_train.reset_index(inplace=True, drop=True)
data_test.reset_index(inplace=True, drop=True)
print('Data for Modeling: ' + str(data_train.shape))
print('Unseen Data For Predictions: ' + str(data_test.shape))

Data for Modeling: (27677, 14)
Unseen Data For Predictions: (4884, 14)


## 5.3 - Understanding the Data

Let's try to understand our dataset now in terms of the given attributes.

We use a dataset modified dataset from UCI called [Adult Data Set
](https://archive.ics.uci.edu/ml/datasets/adult). 

This dataset contains census data and details about various aspects of people and their income.

There are 32561 samples and 14 features. 

Brief descriptions of each column are as follows:

- __age__: continuous; age of the person

- __workclass__: categorical; working class of the person;
Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.

- __education__: categorical; educational qualification of the person;
Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.

- __education-num__: discrete numeric; educational qualification of the person as a encoded value; 

- __marital-status__: categorical; marital status of the person; 
Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.

- __occupation__: categorical; occupation of the person;
Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.

- __relationship__: categorical; relationship information;
Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.

- __race__: categorical; race information;
White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.

- __sex__: categorical; gender of the person; 
Female, Male.

- __capital-gain__: continuous; overall capital gain 

- __capital-loss__: continuous; overall capital loss

- __hours-per-week__: continuous; working hours per week

- __native-country__: categorical; native country of residence;
United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.

- **`income >50K`**: Whether the income of the person is more than $50K (1=yes, 0=no) Target Column

The original dataset and data dictionary can be [found here](https://archive.ics.uci.edu/ml/datasets/adult).

In [ ]:
data_train.head()

age         workclass     education  education-num      marital-status  \
0   27           Private  Some-college             10            Divorced   
1   45         State-gov       HS-grad              9  Married-civ-spouse   
2   29           Private     Bachelors             13  Married-civ-spouse   
3   30           Private     Bachelors             13       Never-married   
4   29  Self-emp-not-inc  Some-college             10            Divorced   

          occupation   relationship   race     sex  capital-gain  \
0       Adm-clerical  Not-in-family  White  Female             0   
1    Exec-managerial           Wife  White  Female             0   
2    Exec-managerial        Husband  Black    Male             0   
3  Machine-op-inspct  Not-in-family  White  Female             0   
4       Craft-repair  Not-in-family  White    Male          2202   

   capital-loss  hours-per-week native-country  income >50K  
0             0              38  United-States            0  
1             0              40  United-States            0  
2             0              55  United-States            1  
3             0              40  United-States            0  
4             0              50  United-States            0

# 6 - PyCaret Environment Setup

The `setup()` function initializes the environment in `pycaret` and creates the transformation pipeline to prepare the data for modeling and deployment. 

`setup()` must be called before executing any other function in `pycaret`. 

It takes two mandatory parameters: a `pandas` dataframe and the name of the target column. 

All other parameters are optional and are used to customize the pre-processing pipeline.

***This is a heavily loaded function because you can use its arguments to do a lot of powerful pre-processing, feature engineering and feature selection steps!***

When `setup()` is executed, PyCaret's inference algorithm will automatically infer the data types for all features based on certain properties. 

The data types are usually inferred correctly but this is not always the case. 

To account for this, PyCaret displays a table containing the features and their inferred data types after `setup()` is executed. 

If all of the data types are correctly identified enter can be pressed to continue or quit can be typed to end the expriment. 

Ensuring that the data types are correct is of fundamental importance in PyCaret as it automatically performs a few pre-processing tasks which are imperative to any machine learning experiment. 

These tasks are performed differently for each data type which means it is very important for them to be correctly configured.

If there are columns which have incorrect indentified data types, we can leverage the `numeric_features` and `categorical_features` parameters in `setup()`.

In [ ]:
from pycaret.classification import *

/usr/local/lib/python3.7/dist-packages/distributed/config.py:20: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  defaults = yaml.load(f)


In [ ]:
experiment = setup(data=data_train, target='income >50K', session_id=42)

Description             Value
0                               session_id                42
1                                   Target       income >50K
2                              Target Type            Binary
3                            Label Encoded              None
4                            Original Data       (27677, 14)
5                           Missing Values              True
6                         Numeric Features                 4
7                     Categorical Features                 9
8                         Ordinal Features             False
9                High Cardinality Features             False
10                 High Cardinality Method              None
11                   Transformed Train Set      (19373, 104)
12                    Transformed Test Set       (8304, 104)
13                      Shuffle Train-Test              True
14                     Stratify Train-Test             False
15                          Fold Generator   StratifiedKFold
16                             Fold Number                10
17                                CPU Jobs                -1
18                                 Use GPU             False
19                          Log Experiment             False
20                         Experiment Name  clf-default-name
21                                     USI              b9ee
22                         Imputation Type            simple
23          Iterative Imputation Iteration              None
24                         Numeric Imputer              mean
25      Iterative Imputation Numeric Model              None
26                     Categorical Imputer          constant
27  Iterative Imputation Categorical Model              None
28           Unknown Categoricals Handling    least_frequent
29                               Normalize             False
30                        Normalize Method              None
31                          Transformation             False
32                   Transformation Method              None
33                                     PCA             False
34                              PCA Method              None
35                          PCA Components              None
36                     Ignore Low Variance             False
37                     Combine Rare Levels             False
38                    Rare Level Threshold              None
39                         Numeric Binning             False
40                         Remove Outliers             False
41                      Outliers Threshold              None
42                Remove Multicollinearity             False
43             Multicollinearity Threshold              None
44             Remove Perfect Collinearity              True
45                              Clustering             False
46                    Clustering Iteration              None
47                     Polynomial Features             False
48                       Polynomial Degree              None
49                    Trignometry Features             False
50                    Polynomial Threshold              None
51                          Group Features             False
52                       Feature Selection             False
53                Feature Selection Method           classic
54            Features Selection Threshold              None
55                     Feature Interaction             False
56                           Feature Ratio             False
57                   Interaction Threshold              None
58                           Fix Imbalance             False
59                    Fix Imbalance Method             SMOTE

Once the setup has been succesfully executed it prints the information grid which contains several important pieces of information. 

Most of the information is related to the pre-processing pipeline which is constructed when `setup()` is executed. 

We are not doing any extensive pre-processing to start with, however a few important things to note at this stage include:

- **session_id :**  A pseudo-random number distributed as a seed in all functions for later reproducibility. If no `session_id` is passed, a random number is automatically generated that is distributed to all functions. In this experiment, the `session_id` is set as `123` for later reproducibility.

- **Target Type :**  Binary or Multiclass. The Target type is automatically detected and shown. There is no difference in how the experiment is performed for Binary or Multiclass problems. All functionalities are identical.

- **Label Encoded :**  When the Target variable is of type string (i.e. 'Yes' or 'No') instead of 1 or 0, it automatically encodes the label into 1 and 0 and displays the mapping (0 : No, 1 : Yes) for reference. In this experiment no label encoding is required since the target variable is of type numeric.

- **Original Data :**  Displays the original shape of the dataset. In this experiment (27677, 14) means 27,677 samples and 14 features including the target column. 

- **Missing Values :**  When there are missing values in the original data this will show as True. For this experiment there are several missing values in the dataset. 

- **Numeric Features :**  The number of features inferred as numeric. In this dataset, 4 features are inferred as numeric. 

- **Categorical Features :**  The number of features inferred as categorical. In this dataset, 9 features are inferred as categorical.

- **Transformed Train Set :**  Displays the shape of the transformed training set. Notice that the original shape of (27677, 24) is transformed into (19373, 104) for the transformed train set and the number of features have increased to 104 due to categorical encoding 

- **Transformed Test Set :**  Displays the shape of the transformed test/hold-out set. There are 8304 samples in test/hold-out set. This split is based on the default value of 70/30 that can be changed using the `train_size` parameter in setup. Can also be used as a validation set if you make decisions of choosing the best model based on this subset.

Notice how a few tasks that are imperative to perform modeling are automatically handled such as missing value imputation, categorical encoding etc. 

Most of the parameters in `setup()` are optional and used for customizing the pre-processing pipeline.



# 7 - Comparing all Models

Comparing all models to evaluate performance is the recommended starting point for modeling once the setup is completed (unless you exactly know what kind of model you need, which is often not the case). 

This function trains all models in the model library and scores them using stratified cross validation for metric evaluation. The output prints a score grid that shows average Accuracy, AUC, Recall, Precision, F1, Kappa, and MCC accross the folds (10 by default) along with training times.

We use a standard 5-fold cross validation

In [ ]:
best_model = compare_models(fold=5)

Model  Accuracy     AUC  Recall   Prec.  \
lightgbm  Light Gradient Boosting Machine    0.8689  0.9236  0.6416  0.7739   
gbc          Gradient Boosting Classifier    0.8623  0.9163  0.5726  0.7964   
ada                  Ada Boost Classifier    0.8573  0.9077  0.6062  0.7516   
lr                    Logistic Regression    0.8470  0.9011  0.5870  0.7231   
rf               Random Forest Classifier    0.8457  0.8869  0.6094  0.7075   
knn                K Neighbors Classifier    0.8393  0.8667  0.6109  0.6855   
lda          Linear Discriminant Analysis    0.8377  0.8891  0.5539  0.7063   
ridge                    Ridge Classifier    0.8368  0.0000  0.4969  0.7378   
et                 Extra Trees Classifier    0.8267  0.8431  0.5855  0.6559   
dt               Decision Tree Classifier    0.8149  0.7578  0.6094  0.6159   
nb                            Naive Bayes    0.7970  0.8913  0.8144  0.5523   
svm                   SVM - Linear Kernel    0.7964  0.0000  0.4840  0.5926   
dummy                    Dummy Classifier    0.7600  0.5000  0.0000  0.0000   
qda       Quadratic Discriminant Analysis    0.2462  0.5033  0.9976  0.2412   

              F1   Kappa     MCC  TT (Sec)  
lightgbm  0.7014  0.6184  0.6230     0.380  
gbc       0.6661  0.5823  0.5949     2.858  
ada       0.6709  0.5812  0.5868     0.918  
lr        0.6480  0.5515  0.5564     4.788  
rf        0.6546  0.5560  0.5587     2.414  
knn       0.6459  0.5424  0.5440     4.828  
lda       0.6208  0.5195  0.5257     0.424  
ridge     0.5937  0.4965  0.5117     0.074  
et        0.6185  0.5068  0.5083     2.978  
dt        0.6124  0.4909  0.4911     0.192  
nb        0.6581  0.5212  0.5411     0.078  
svm       0.5280  0.4012  0.4072     0.198  
dummy     0.0000  0.0000  0.0000     0.052  
qda       0.3885  0.0032  0.0330     0.260

Just one line of code has trained and evaluated around 14 models using cross validation. 

The score grid printed above by default is sorted using 'Accuracy' (highest to lowest) which can be changed by passing the `sort` parameter. 

For example `compare_models(sort = 'Recall')` will sort the grid by Recall instead of Accuracy. 

If you want to change the fold parameter from the default value of `10` to a different value then you can use the `fold` parameter. For example `compare_models(fold = 5)` will compare all models on 5 fold cross validation as we did above. 

Reducing the number of folds will improve the training time. By default, `compare_models` return the best performing model based on default sort order but can be used to return a list of top N models by using `n_select` parameter.

In [ ]:
best_model

LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
               importance_type='split', learning_rate=0.1, max_depth=-1,
               min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
               n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
               random_state=42, reg_alpha=0.0, reg_lambda=0.0, silent='warn',
               subsample=1.0, subsample_for_bin=200000, subsample_freq=0)

Looks like boosting models have taken up the leaderboard above!

# 8 - Create ML Models

`create_model` is one of the most important functions in PyCaret and is often the starting point or foundation behind most of the PyCaret functionalities. 

As the name suggests this function trains and evaluates a model using cross validation that can be set with `fold` parameter. The output prints a score grid that shows Accuracy, AUC, Recall, Precision, F1, Kappa and MCC by fold. 

For the remaining part of this tutorial, we will work with the following two models as our candidate models. 

The selections are for illustration purposes only where we pick the best model which is a boosting model and a simple logistic model just to compare:

```
- Light Gradient Boosting Machine ('lightgbm')
- Logistic Regression ('lr')
```

There are 18 classifiers available in the model library of PyCaret. To see list of all classifiers either check the `docstring` or use `models` function to see the library.

In [ ]:
models()

Name  \
ID                                          
lr                    Logistic Regression   
knn                K Neighbors Classifier   
nb                            Naive Bayes   
dt               Decision Tree Classifier   
svm                   SVM - Linear Kernel   
rbfsvm                SVM - Radial Kernel   
gpc           Gaussian Process Classifier   
mlp                        MLP Classifier   
ridge                    Ridge Classifier   
rf               Random Forest Classifier   
qda       Quadratic Discriminant Analysis   
ada                  Ada Boost Classifier   
gbc          Gradient Boosting Classifier   
lda          Linear Discriminant Analysis   
et                 Extra Trees Classifier   
lightgbm  Light Gradient Boosting Machine   
dummy                    Dummy Classifier   

                                                  Reference  Turbo  
ID                                                                  
lr        sklearn.linear_model._logistic.LogisticRegression   True  
knn       sklearn.neighbors._classification.KNeighborsCl...   True  
nb                           sklearn.naive_bayes.GaussianNB   True  
dt             sklearn.tree._classes.DecisionTreeClassifier   True  
svm       sklearn.linear_model._stochastic_gradient.SGDC...   True  
rbfsvm                             sklearn.svm._classes.SVC  False  
gpc       sklearn.gaussian_process._gpc.GaussianProcessC...  False  
mlp       sklearn.neural_network._multilayer_perceptron....  False  
ridge           sklearn.linear_model._ridge.RidgeClassifier   True  
rf          sklearn.ensemble._forest.RandomForestClassifier   True  
qda       sklearn.discriminant_analysis.QuadraticDiscrim...   True  
ada       sklearn.ensemble._weight_boosting.AdaBoostClas...   True  
gbc         sklearn.ensemble._gb.GradientBoostingClassifier   True  
lda       sklearn.discriminant_analysis.LinearDiscrimina...   True  
et            sklearn.ensemble._forest.ExtraTreesClassifier   True  
lightgbm                    lightgbm.sklearn.LGBMClassifier   True  
dummy                         sklearn.dummy.DummyClassifier   True

## 8.1 - Create Light Gradient Boosting Model

In [ ]:
lgbm = create_model('lightgbm', fold=5)

Accuracy     AUC  Recall   Prec.      F1   Kappa     MCC
Fold                                                          
0       0.8663  0.9205  0.6344  0.7682  0.6949  0.6103  0.6149
1       0.8712  0.9289  0.6462  0.7795  0.7066  0.6251  0.6296
2       0.8650  0.9236  0.6516  0.7528  0.6986  0.6122  0.6148
3       0.8699  0.9209  0.6226  0.7910  0.6968  0.6154  0.6226
4       0.8722  0.9240  0.6534  0.7782  0.7104  0.6292  0.6331
Mean    0.8689  0.9236  0.6416  0.7739  0.7014  0.6184  0.6230
Std     0.0028  0.0030  0.0116  0.0128  0.0060  0.0074  0.0075

## 8.2 - Create Logistic Regression Model

# 9 - Tune ML Models

When a model is created using the `create_model()` function it uses the default hyperparameters to train the model. 

In order to tune hyperparameters, the [`tune_model()`](https://pycaret.readthedocs.io/en/latest/api/classification.html#pycaret.classification.tune_model) function is used. 

This function automatically tunes the hyperparameters of a model using `Randomized Search` on a pre-defined search space. The output prints a score grid that shows Accuracy, AUC, Recall, Precision, F1, Kappa, and MCC by fold for the best model. 

To use the custom search grid, you can pass `custom_grid` parameter in the `tune_model` function (see 9.2 Logistic Regression tuning below).

## 9.1 Tune Light Gradient Boosting Model

## 9.2 Tune Logistic Regression Model

By default, `tune_model` optimizes `Accuracy` but this can be changed using `optimize` parameter. 

For example: `tune_model(lgbm, optimize = 'AUC')` will search for the hyperparameters of a Light Gradient Boosting Classifier that results in the highest `AUC` instead of `Accuracy`. 

For the purposes of this example, we have used the default metric `Accuracy` only for the sake of simplicity. 

Generally, when the dataset is imbalanced (such as the adult dataset we are working with) `Accuracy` is not a good metric for consideration. 

Metrics alone are not the only criteria you should consider when finalizing the best model for production. Other factors to consider include training time, standard deviation of kfolds etc. 

For now, let's move forward considering the Tuned Light Gradient Boosting Classifier `tuned_lgbm`, as our best model for the remainder of this tutorial.

# 10 - Plot ML Model Evaluation Diagnostics

Before model finalization, the `plot_model()` function can be used to analyze and evaluate the model performance across different aspects such as AUC, confusion_matrix, decision boundary etc. 

This function takes a trained model object and returns a plot based on the test / hold-out set. 

There are many different plots available, please see the `plot_model()` docstring for the list of available plots.

## 10.1 - Confusion Matrix

## 10.2 - Feature Importance

*Another* way to analyze the performance of models is to use the `evaluate_model()` function which displays a user interface for all of the available plots for a given model. It internally uses the `plot_model()` function. 

# 11 - Predict on Test / Hold-out Sample

Before finalizing the model, it is advisable to perform one final check by predicting on the test/hold-out set and reviewing the evaluation metrics. 

If you look at the information in Section 6 above, you will see that 30% (8304 samples) of the data has been separated out as test/hold-out sample. 

All of the evaluation metrics we have seen above are cross validated results based on the training set (70%) only. The `plot_model` shows results on the Test Data.

We can use our trained model stored in the `tuned_lgbm` variable to get predictions from our test data and use them as necessary

The accuracy on test/hold-out set is **`0.8746`** compared to **`0.8701`** achieved on the `tuned_lgbm` CV results earlier. 

This is not a significant difference. If there is a large variation between the test/hold-out and CV results, then this would normally indicate over-fitting but could also be due to several other factors and would require further investigation. 

In this case, we will move forward with finalizing the model and predicting on unseen data (the 15% that we had separated in the beginning and never exposed to PyCaret experiments).

# 12 - Finalize Model for Deployment

Model finalization is the last step in the experiment. 

A normal machine learning workflow in PyCaret starts with `setup()`, followed by comparing all models using `compare_models()` and shortlisting a few candidate models (based on the metric of interest) to perform several modeling techniques such as hyperparameter tuning, ensembling, stacking etc. (more on advanced techniques in the next tutorial!).

This workflow will eventually lead you to the best model for use in making predictions on new and unseen data. 


The `finalize_model()` function fits the model onto the complete dataset **including** the test/hold-out sample (30% in this case). The purpose of this function is to **train the model on the complete dataset** before it is deployed in production.

In [ ]:
#Final Light Gradient Boosting Model to be used for deployment


**Caution:** Once the model is finalized using `finalize_model()`, the entire dataset including the test/hold-out set is used for training. 

As such, if the model is used for predictions on the hold-out set after `finalize_model()` is used, the results will be misleading as you are trying to predict on the same data that was used for modeling. 

In order to demonstrate this point only, we will use `final_lgbm` under `predict_model()` to compare the information grid with the one above in section 11. 

You can clearly see even if the model is the same, because it was also trained on the test data one final time, accuracy is now higher at __0.8886__ as compared to our model in Section 11 with __0.8746__ accuracy.

# 13 - Predict on unseen / new datasets

The `predict_model()` function is also used to predict on any new / unseen datasets. 

The only difference from section 11 above is that this time we will pass the `data_test` parameter. `data_test` is the variable created at the beginning of the tutorial and contains 15% of the original dataset which was never exposed to PyCaret. (see section 5 for explanation)

The `Label` and `Score` columns are added onto the `data_test` set. 

Label is the prediction and score is the probability of the prediction. 

Notice that predicted results are concatenated to the original dataset while all the data transformations are automatically performed in the background. 


# 14 - Saving your ML Model

We have now finished the experiment by finalizing the `tuned_lgbm` model which is now stored in `final_lgbm` variable. 

We have also used the model stored in `final_lgbm` to predict `data_test`. 

This brings us to the end of our experiment, but one question is still to be asked: What happens when you have more new data to predict? Do you have to go through the entire experiment again? 

The answer is no, PyCaret's inbuilt function `save_model()` allows you to save the model along with entire transformation pipeline for later use.

# 15 - Load Saved ML Model for future use

To load a saved model at a future date in the same or an alternative environment, we would use PyCaret's `load_model()` function and then easily apply the saved model on new unseen data for prediction.